File: 139-word-break/word-break.cpp
class Solution {
public:
    unordered_set<string>st;
    int n;
    vector<int>dp;



    bool  solve(int i , string &s){
        if(i==n)return true;

        if(dp[i]!=-1)return dp[i];
        
        for(int j = i ; j < n ; j++)
        {
            if(st.count(s.substr(i,j-i+1))){
                bool found_right=solve(j+1,s);
                if(found_right==true)return true;
            }
        }
        return dp[i]=false;

    }

    
    bool wordBreak(string s, vector<string>& wordDict) {

        n=s.size();
        dp.assign(n+1,-1);

        for(auto &word:wordDict){
            st.insert(word);
        }
        return solve(0,s);




    }
};

/*
    leetcode 
    l  solve("eetcode")
    le + solv("etcode")
    .................
    ..............
    leet + solve("code")---> here we found substring (0, 3) && solve "code" so can return true; 
    else return false;
*/

File: arrays-and-strings/139-word-break/word-break.java
class Solution {
    int n;
    Set<String>seen;
    Boolean dp[];
    boolean  solve(int i,String s){
        if (i==n)return true;
        if(dp[i] != null )return dp[i];
        for(int  j  = i ; j < n ; j++){
            if(seen.contains(s.substring(i,j+1))){
                boolean found_right= solve(j+1,s);
                if(found_right){
                    return dp[i] = true;
                }
            }

        }
        return  dp[i] = false;
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        //word break
        //all possible breaks;
        //applepenapple , apple+pen:
        dp = new Boolean[301];
        n = s.length();
        seen = new HashSet<>();
        for(String word : wordDict){
            seen.add(word);
        }
        return solve(0,s);



        
        
    }
}